crates/bevy_utils/src/lib.rs

```diff
@@ -221,3 +221,197 @@ impl<K: Hash + Eq + PartialEq + Clone, V> PreHashMapExt<K, V> for PreHashMap<K,
+    T: Send + Sync + Copy + Add + AddAssign + Sub + SubAssign,
+{
+    /// The minimum value that the progress can have, inclusive.
+    min: T,
+    /// The maximum value that the progress can have, inlucsive.
+    max: T,
+    /// The current value of progress.
+    value: T,
+}
+
+impl<T: Send + Sync + Copy + Add + AddAssign + Sub + SubAssign> Progress<T>
+where
+    T: PartialOrd<T>,
+{
+    /// Creates a new progress using a `value`, and a `min` and `max` that defines a `range`.
+    ///
+    /// The `value` must be within the bounds of the `range` or returns a [`ProgressError`].
+    pub fn new(value: T, min: T, max: T) -> Result<Self, ProgressError> {
+        Self::from_range(value, min..=max)
+    }
+
+    /// Creates a new progress using a `value` and a `range`.
+    ///
+    /// The `value` must be within the bounds of the `range` or returns a [`ProgressError::OutOfBounds`].
+    pub fn from_range(value: T, range: RangeInclusive<T>) -> Result<Self, ProgressError> {
+        if range.start() >= range.end() {
+            Err(ProgressError::InvalidRange)
+        } else if range.contains(&value) {
+            Ok(Self {
+                value,
+                min: *range.start(),
+                max: *range.end(),
+            })
+        } else {
+            Err(ProgressError::OutOfBounds)
+        }
+    }
+
+    /// Gets the min bound of the progress.
+    pub fn min(&self) -> T {
+        self.min
+    }
+
+    /// Gets the max bound of the progress.
+    pub fn max(&self) -> T {
+        self.max
+    }
+
+    /// Gets the bounds of the progress.
+    pub fn bounds(&self) -> RangeInclusive<T> {
+        self.min..=self.max
+    }
+
+    /// Gets the current value of progress.
+    pub fn progress(&self) -> T {
+        self.value
+    }
+
+    /// Sets the progress to a new value and returns the new value if successful.
+    ///
+    /// The `value` must be within the bounds of the `range` or returns a [`ProgressError::OutOfBounds`].
+    pub fn set_progress(&mut self, new_value: T) -> Result<T, ProgressError> {
+        if self.bounds().contains(&new_value) {
+            self.value = new_value;
+            Ok(self.value)
+        } else {
+            Err(ProgressError::OutOfBounds)
+        }
+    }
+}
+
+impl Progress<f32> {
+    /// Creates a new [`Progress`] using percent.
+    /// `Min` = 0.0
+    /// `Max` = 100.0
+    pub fn from_percent(value: f32) -> Self {
+        Self::from_range(value, 0.0..=100.0).unwrap()
+    }
+
+    /// Returns the current progress, normalized between 0 and 1.
+    ///
+    /// 0 represents value == min,
+    /// 1 represents value == max.
+    pub fn normalized(&self) -> f32 {
+        remap_range(self.value, (self.min, self.max), (0.0, 1.0))
+    }
+}
+
+impl Default for Progress<f32> {
+    fn default() -> Self {
+        Self {
+            min: 0.0,
+            max: 1.0,
+            value: 0.0,
+        }
+    }
+}
+
+/// Error types for [`Progress`].
+#[derive(Debug, PartialEq)]
```

Comment:
I think this should use `thiserror` crate for custom error messages.

Reply:
Agreed